k2/csrc/multisplit.h

```diff
@@ -0,0 +1,191 @@
+ *                          it will contain a sorted version of `in_values`.
+ */
+template <typename Lambda>
+void MultiSplitKeyValuePairs(ContextPtr context, int32_t num_elements,
+                             int32_t num_buckets, Lambda *bucket_mapping_func,
+                             const uint32_t *in_keys, const uint32_t *in_values,
+                             uint32_t *out_keys, uint32_t *out_values) {
+  K2_CHECK_GT(num_elements, 0);
+  K2_CHECK_GT(num_buckets, 0);
+  int32_t log_buckets = static_cast<int32_t>(ceilf(log2f(num_buckets)));
+  cudaStream_t stream = context->GetCudaStream();
+
+  constexpr int32_t kMultiSplitNumWarps = 8;
+  int32_t num_threads = kMultiSplitNumWarps * 32;
+  int32_t num_blocks = (num_elements + num_threads - 1) / num_threads;
+
+  Array1<uint32_t> d_mask(context, num_elements, 0);
+  const uint32_t *mask_data = in_keys;
+
+  if (bucket_mapping_func != nullptr) {
+    K2_CUDA_SAFE_CALL(MarkBinsGeneral<<<num_blocks, num_threads, 0, stream>>>(
+        in_keys, num_elements, *bucket_mapping_func, d_mask.Data()));
+    mask_data = d_mask.Data();
+  }
+
+  Array1<uint64_t> d_key_value_pairs(context, num_elements);
+  K2_CUDA_SAFE_CALL(
+      PackingKeyValuePairs<<<num_blocks, num_threads, 0, stream>>>(
+          num_elements, in_keys, in_values, d_key_value_pairs.Data()));
+
+  Array1<uint32_t> d_out(context, num_elements, 0);
+  size_t temp_storage_bytes = 0;
+
+  K2_CUDA_SAFE_CALL(cub::DeviceRadixSort::SortPairs(
+      nullptr, temp_storage_bytes, mask_data, d_out.Data(),
+      d_key_value_pairs.Data(), d_key_value_pairs.Data(), num_elements, 0,
+      log_buckets, stream));
+
+  Array1<int8_t> d_temp_storage(context, temp_storage_bytes);
+
+  K2_CUDA_SAFE_CALL(cub::DeviceRadixSort::SortPairs(
+      d_temp_storage.Data(), temp_storage_bytes, mask_data, d_out.Data(),
+      d_key_value_pairs.Data(), d_key_value_pairs.Data(), num_elements, 0,
+      log_buckets, stream));
+
+  K2_CUDA_SAFE_CALL(
+      UnpackingKeyValuePairs<<<num_blocks, num_threads, 0, stream>>>(
+          num_elements, d_key_value_pairs.Data(), out_keys, out_values));
+}
+
+// See the doc for `MultiSplitKeyValuePairs`.
+// The only difference is that this function accepts only `keys`, no `values`.
+template <typename Lambda>
+void MultiSplitKeysOnly(ContextPtr context, int32_t num_elements,
+                        int32_t num_buckets, Lambda *bucket_mapping_func,
+                        const uint32_t *in_keys, uint32_t *out_keys) {
+  K2_CHECK_GT(num_elements, 0);
+  K2_CHECK_GT(num_buckets, 0);
+  int32_t log_buckets = static_cast<int32_t>(ceilf(log2f(num_buckets)));
+  cudaStream_t stream = context->GetCudaStream();
+
+  constexpr int32_t kMultiSplitNumWarps = 8;
+  int32_t num_threads = kMultiSplitNumWarps * 32;
+  int32_t num_blocks = (num_elements + num_threads - 1) / num_threads;
+
+  Array1<uint32_t> d_mask(context, num_elements, 0);
+  const uint32_t *mask_data = in_keys;
+
+  if (bucket_mapping_func != nullptr) {
```

Comment:
I would suggest we declare `d_mask` in the `if` branch so that we don't do memory allocation when `bucket_mapping_func`. you can do as below:
```
Array<uint32_t> d_mask;
if(bucket_mapping_func != nullptr){
d_mask = Array1<unint32_t>(context, num_elements);
....
}
```